packages/wrangler/templates/startDevWorker/InspectorProxyWorker.ts

```diff
@@ -0,0 +1,472 @@
+// TODO: Inspector Durable Object
+//  1. ~~Durable Object websocket server~~
+//  2. ~~JSON list endpoints (proxy through to workerd and filter for core:user:*)~~
+//  3. ~~Receive cf data from proxy controller (use env.SECRET for auth)~~
+//  4. ~~Connect to remote websocket and establish devtools connection~~
+//  5. ~~Buffer messages from remote until we've got client connection~~
+//  6. ~~Intercept logging messages and log them to the console (probably via websocket to proxy controller)~~
+//  7. Rewriting messages for source maps and stuff
+//
+
+/**
+ * TODO:
+ *
+ * - Rewriting messages for source maps and stuff
+ *      -
+ * - tests
+ *      - test devtools get-source-map triggers service binding callback?
+ *      -
+ */
+
+import assert from "node:assert";
+import {
+	castErrorCause,
+	DevToolsCommandRequest,
+	DevToolsCommandRequests,
+	DevToolsCommandResponse,
+	DevToolsCommandResponses,
+	DevToolsEvent,
+	DevToolsEvents,
+	type InspectorProxyWorkerIncomingWebSocketMessage,
+	type InspectorProxyWorkerOutgoingRequestBody,
+	type InspectorProxyWorkerOutgoingWebsocketMessage,
+	type ProxyData,
+} from "../../src/api/startDevWorker/events";
+import {
+	MaybePromise,
+	assertNever,
+	createDeferredPromise,
+} from "../../src/api/startDevWorker/utils";
+
+interface Env {
+	PROXY_CONTROLLER: Fetcher;
+	PROXY_CONTROLLER_AUTH_SECRET: string;
+	WRANGLER_VERSION: string;
+	DURABLE_OBJECT: DurableObjectNamespace;
+}
+
+export default {
+	fetch(req, env) {
+		const singleton = env.DURABLE_OBJECT.idFromName("");
+		const inspectorProxy = env.DURABLE_OBJECT.get(singleton);
+
+		return inspectorProxy.fetch(req);
+	},
+} as ExportedHandler<Env>;
+
+export class InspectorProxyWorker implements DurableObject {
+	constructor(_state: DurableObjectState, readonly env: Env) {}
+
+	#websockets: {
+		proxyController?: WebSocket;
+		runtime?: WebSocket;
+		devtools?: WebSocket;
+	} = {};
+	#runtimeWebSocketPromise = createDeferredPromise<WebSocket>();
+
+	#proxyData?: ProxyData;
+
+	runtimeMessageBuffer: string[] = [];
+
+	#handleRuntimeIncomingMessage = (event: MessageEvent) => {
+		assert(typeof event.data === "string");
+		assert(
+			this.#websockets.proxyController !== undefined,
+			"Expected proxy controller websocket"
+		);
+
+		const msg = JSON.parse(event.data) as
+			| DevToolsCommandResponses
+			| DevToolsEvents;
+		this.sendDebugLog("RUNTIME INCOMING MESSAGE", msg);
+
+		if (
+			"method" in msg &&
+			(msg.method === "Runtime.exceptionThrown" ||
+				msg.method === "Runtime.consoleAPICalled")
+		) {
+			this.sendProxyControllerMessage(event.data);
+		}
+
+		this.runtimeMessageBuffer.push(JSON.stringify(msg));
```

Comment:
TODO we don't need to stringify every message

This is a remnant of trying to rewrite some messages to fix a bug mentioned in another comment

```suggestion
		this.runtimeMessageBuffer.push(event.data);
```

Reply:
We're gonna keep the parsed objects rather than the strings just because we need to intercept and modify and we may want to do more in the future

Memory-wise, we can see if this makes any difference at a later date